gmt/gridding.py

```diff
@@ -0,0 +1,77 @@
+"""
+GMT modules for Gridding of Data Tables
+"""
+import xarray as xr
+
+from .clib import Session
+from .helpers import (
+    build_arg_string,
+    fmt_docstring,
+    GMTTempFile,
+    use_alias,
+    data_kind,
+    dummy_context,
+)
+from .exceptions import GMTInvalidInput
+
+
+@fmt_docstring
+@use_alias(I="spacing", R="region")
+def surface(x=None, y=None, z=None, data=None, **kwargs):
+    """
+    Grids table data using adjustable tension continuous curvature splines
+
+    Takes a matrix, (x,y,z) pairs, or a file name as input.
+
+    Must provide either *data* or *x*, *y*, and *z*.
+
+    {gmt_module_docs}
+
+    Parameters
+    ----------
+    x, y, z : 1d arrays
+        Arrays of x and y coordinates and values z of the data points.
+    data : str or 2d array
+        Either a data file name or a 2d numpy array with the tabular data.
+
+    spacing (I) :
+        ``'xinc[unit][+e|n][/yinc[unit][+e|n]]'``.
+        x_inc [and optionally y_inc] is the grid spacing.
+
+    region (R) : str or list
+        ``'xmin/xmax/ymin/ymax[+r][+uunit]'``.
+        Specify the region of interest.
+
+    outfile (G) : str
+        Optional. The file name for the output netcdf file with extension .nc
+        to store the grid in.
+
+    {aliases}
+
+    Returns
+    -------
+    array: xarray.DataArray
+        The output grid as a DataArray
+    """
+    kind = data_kind(data, x, y, z)
+    if kind == "vectors" and z is None:
+        raise GMTInvalidInput("Must provide z with x and y.")
+
+    with GMTTempFile(suffix=".nc") as tmpfile:
+        with Session() as lib:
+            if kind == "file":
+                file_context = dummy_context(data)
+            elif kind == "matrix":
+                file_context = lib.virtualfile_from_matrix(data)
+            elif kind == "vectors":
+                file_context = lib.virtualfile_from_vectors(x, y, z)
+            else:
+                raise GMTInvalidInput("Unrecognized data type: {}".format(type(data)))
+            with file_context as infile:
+                if "G" not in kwargs.keys():
+                    kwargs.update({"G": tmpfile.name})
+                outfile = kwargs["G"]
+                arg_str = " ".join([infile, build_arg_string(kwargs)])
+                lib.call_module(module="surface", args=arg_str)
+        result = xr.open_dataset(outfile)
```

Comment:
```suggestion
        result = xr.open_dataset(outfile).load()
        result.close()
```

xarray sometimes doesn't really load the data into memory, it creates a memory map to the file on disk. Since this is a temporary file, we have to make sure it's loaded into memory and the file is closed.

Reply:
Let me try testing `xr.open_dataset` on a big dataset (>10GB) I have first. I feel like this should be wrapped around a python `with` statement to make sure the file does close even if it's half loaded or something.

Reply:
Ok, I've forced the data to be loaded/persisted, using:

```python
    with xr.open_dataset(outfile) as dataset:
        result = dataset.load()
```

Ideally there would be a big dataset we can add to our test suite to make sure this works, but it might slow things down too much so I'll leave it at that.